src/etl/impl/SubscriptionSource.cpp

```diff
@@ -0,0 +1,336 @@
+        static constexpr char const* const JS_Type = "type";
+        static constexpr char const* const JS_Transaction = "transaction";
+        static constexpr char const* const JS_ValidationReceived = "validationReceived";
+        static constexpr char const* const JS_ManifestReceived = "manifestReceived";
+
+        if (object.contains(JS_Result)) {
+            auto const& result = object.at(JS_Result).as_object();
+            if (result.contains("ledger_index"))
+                ledgerIndex = result.at(JS_LedgerIndex).as_int64();
+
+            if (result.contains(JS_ValidatedLedgers)) {
+                auto validatedLedgers = boost::json::value_to<std::string>(result.at(JS_ValidatedLedgers));
+                setValidatedRange(std::move(validatedLedgers));
+            }
+            LOG(log_.info()) << "Received a message on ledger subscription stream. Message : " << object;
+
+        } else if (object.contains(JS_Type) && object.at(JS_Type) == JS_LedgerClosed) {
+            LOG(log_.info()) << "Received a message on ledger subscription stream. Message : " << object;
+            if (object.contains(JS_LedgerIndex)) {
+                ledgerIndex = object.at(JS_LedgerIndex).as_int64();
+            }
+            if (object.contains(JS_ValidatedLedgers)) {
+                auto validatedLedgers = boost::json::value_to<std::string>(object.at(JS_ValidatedLedgers));
+                setValidatedRange(std::move(validatedLedgers));
+            }
+
+        } else {
+            if (isForwarding_) {
+                if (object.contains(JS_Transaction)) {
+                    dependencies_.forwardProposedTransaction(object);
+                } else if (object.contains(JS_Type) && object.at(JS_Type) == JS_ValidationReceived) {
+                    dependencies_.forwardValidation(object);
+                } else if (object.contains(JS_Type) && object.at(JS_Type) == JS_ManifestReceived) {
+                    dependencies_.forwardManifest(object);
+                }
+            }
+        }
+
+        if (ledgerIndex != 0) {
+            LOG(log_.trace()) << "Pushing ledger sequence = " << ledgerIndex;
+            dependencies_.pushValidatedLedger(ledgerIndex);
+        }
+
+        return std::nullopt;
+    } catch (std::exception const& e) {
+        LOG(log_.error()) << "Exception in handleMessage : " << e.what();
+        return util::requests::RequestError{fmt::format("Error handling message: {}", e.what())};
+    }
+}
+
+void
+SubscriptionSource::handleError(util::requests::RequestError const& error, boost::asio::yield_context yield)
+{
+    isConnected_ = false;
+    if (not stop_) {
+        onDisconnect_();
+        isForwarding_ = false;
+    }
+
+    if (wsConnection_ != nullptr) {
+        auto const error = wsConnection_->close(yield);
+        if (error) {
+            LOG(log_.error()) << "Error closing websocket connection: " << error->message();
+        }
+        wsConnection_.reset();
```

Comment:
Does here have data race problem with the reset in destructor?

Reply:
This happens in coroutine and in the destructor we wait for all async operations to complete before doing reset, so I think there is no data race. But if you see some other dangerous case please describe it.

Reply:
The coroutine posted at line 66 can be executed after wsConnection_. reset in handlerError,  right?  Both are coroutine. The one in handlerError gets executed first.

Reply:
After a test I confirmed that the future is not the last task when using boost coroutines. I will change it to use `boost::asio::use_future`.